beacon.go

```diff
@@ -17,110 +17,185 @@
+	}
+	return nil
+}
+
+func parseInput(r *http.Request) (string, string, int64, error) {
+	refName := r.FormValue("chromosome")
+	if refName == "" {
+		return "", "", 0, newInvalidInputError("parsing chromosome name", errors.New("value is required"))
 	}
-	genomicsService, err := genomics.New(client)
+	allele := r.FormValue("allele")
+	if refName == "" {
+		return "", "", 0, newInvalidInputError("parsing allele name", errors.New("value is required"))
+	}
+	coord, err := strconv.ParseInt(r.FormValue("coordinate"), 10, 64)
 	if err != nil {
-		http.Error(w, "Invalid server configuration", http.StatusInternalServerError)
+		return "", "", 0, newInvalidInputError("parsing coordinate", err)
 	}
-	variantsService := genomics.NewVariantsService(genomicsService)
+	return refName, allele, coord, nil
+}
 
+func writeResponse(w http.ResponseWriter, exists bool) error {
 	type beaconResponse struct {
 		XMLName struct{} `xml:"BEACONResponse"`
 		Exists  bool     `xml:"exists"`
 	}
 	var resp beaconResponse
-
-	for {
-		searchResponse, err := variantsService.Search(search).Do()
-		if err != nil {
-			http.Error(w, err.Error(), http.StatusInternalServerError)
-			return
-		}
-
-		for _, variant := range searchResponse.Variants {
-			if search.Start != variant.Start {
-				continue
-			}
-			if allele == variant.ReferenceBases {
-				resp.Exists = true
-			} else {
-				for _, base := range variant.AlternateBases {
-					if base == allele {
-						resp.Exists = true
-						break
-					}
-				}
-			}
-		}
-
-		if resp.Exists || searchResponse.NextPageToken == "" {
-			break
-		}
-		search.PageToken = searchResponse.NextPageToken
-	}
+	resp.Exists = exists
 
 	w.Header().Set("Content-Type", "application/xml")
 	enc := xml.NewEncoder(w)
 	enc.Indent("", "  ")
-	if err = enc.Encode(resp); err != nil {
-		http.Error(w, "Failed writing response", http.StatusInternalServerError)
+	if err := enc.Encode(resp); err != nil {
+		return newApiError("Serializationerror", http.StatusInternalServerError, "Serializing response", err)
 	}
+	return nil
+}
+
+// apiError is used to capture errors that have been defined in the API.
+type apiError struct {
+	name  string
+	code  int
+	cause error
+}
+
+func (err *apiError) Error() string {
+	return fmt.Sprintf("%s (%d): %v", err.name, err.code, err.cause)
+}
+
+func newApiError(name string, code int, context string, err error) error {
+	return &apiError{name, code, fmt.Errorf("%s: %v", context, err)}
+}
+
+func newInvalidConfigError(context string, err error) error {
+	return newApiError("InvalidConfig", http.StatusPreconditionFailed, context, err)
+}
+
+func newInvalidInputError(context string, err error) error {
+	return newApiError("InvalidInput", http.StatusBadRequest, context, err)
+}
+
+func newDataAccessError(context string, err error) error {
+	return newApiError("DataAccessError", http.StatusInternalServerError, context, err)
+}
+
+// writeError writes either a JSON object or bare HTTP error describing err to
```

Comment:
The JSON object encoding an error is specific to htsget and doesn't apply here.

It looks like Beacon does define a BeaconError that is similar, and we could write one, but that is outside the scope of this change.

Reply:
done